src/SimplifyCorrelatedDifferences.cpp

```diff
@@ -94,9 +94,9 @@ class SimplifyCorrelatedDifferences : public IRMutator {
             (ma == Monotonic::Decreasing && mb == Monotonic::Increasing && std::is_same<T, Add>::value)) {
 
             for (auto it = lets.rbegin(); it != lets.rend(); it++) {
-                if (expr_uses_var(e, it->first)) {
-                    e = Let::make(it->first, it->second, e);
-                }
+                // We must avoid making let nodes here, because CSE
+                // does not handle IR with shadowed variable names.
+                e = graph_substitute(it->first, it->second, e);
```

Comment:
If CSE cannot handle shadowed names, perhaps we should add a visitor to it to verify that it doesn't encounter duplicate lets?

Reply:
Or just fix whatever is broken when there are repeated names. I think this is the minimal PR to unbreak master though, so I'll work on that in a separate PR.